@potiuk  The CI job is failing due to pylint in a file I didn't touch: 

    tests/jobs/test_scheduler_job.py:3663:15: E1101: Instance of 'str' has no 'task_id' member (no-member)

I'm rebased to the latest `origin/branch`  commit 2833c2b2ffda682cc11893bc5efb1f53a254097e

--- a/airflow/models/taskinstance.py
+++ b/airflow/models/taskinstance.py
@@ -44,6 +44,7 @@ from airflow.configuration import conf
 from airflow.exceptions import (
     AirflowException,
     AirflowFailException,
+    AirflowNotFoundException,
     AirflowRescheduleException,
     AirflowSensorTimeout,
     AirflowSkipException,
@@ -51,6 +52,7 @@ from airflow.exceptions import (
     AirflowTaskTimeout,
 )
 from airflow.models.base import COLLATION_ARGS, ID_LEN, Base
+from airflow.models.connection import Connection
 from airflow.models.log import Log
 from airflow.models.taskfail import TaskFail
 from airflow.models.taskreschedule import TaskReschedule
@@ -1637,6 +1639,30 @@ class TaskInstance(Base, LoggingMixin):
                 """Get Airflow Variable after deserializing JSON value"""
                 return Variable.get(item, default_var=default_var, deserialize_json=True)
 
+        class ConnectionAccessor:
+            """
+            Wrapper around Connection. This way you can get connections in
+            templates by using ``{{ conn.value.conn_id }}`` or
+            ``{{ conn.value.get('conn_id', 'fallback') }}``.
+            """
+
+            def __getattr__(
+                self,
+                item: str,
+            ):
+                return Connection.get_connection_from_secrets(item)
+
+            @staticmethod
+            def get(
+                item: str,
+                default_conn: Any = None,
+            ):
+                """Get Airflow Connection value"""
+                try:
+                    return Connection.get_connection_from_secrets(item)
+                except AirflowNotFoundException:
+                    return default_conn
+
         return {
             'conf': conf,
             'dag': task.dag,
@@ -1674,6 +1700,9 @@ class TaskInstance(Base, LoggingMixin):
             'var': {
                 'json': VariableJsonAccessor(),
                 'value': VariableAccessor(),
+            },
+            'conn': {
+                'value': ConnectionAccessor(),
             },
             'yesterday_ds': yesterday_ds,
             'yesterday_ds_nodash': yesterday_ds_nodash,

--- a/docs/apache-airflow/concepts/connections.rst
+++ b/docs/apache-airflow/concepts/connections.rst
@@ -27,7 +27,7 @@ They can be managed via the UI or via the CLI; see :doc:`/howto/connection` for 
 You can use Connections directly from your own code, you can use them via Hooks or use them from :ref:`templates <concepts:jinja-templating>`::
 
 
-    echo {{ conn.<variable_name>.host }}
+    echo {{ conn.<conn_id>.host }}
 
 Hooks
 -----

--- a/docs/apache-airflow/concepts/connections.rst
+++ b/docs/apache-airflow/concepts/connections.rst
@@ -24,8 +24,10 @@ A Connection is essentially set of parameters - such as username, password and h
 
 They can be managed via the UI or via the CLI; see :doc:`/howto/connection` for more information on creating, editing and managing connections. There are customizable connection storage and backend options.
 
-You can use Connections directly from your own code, or you can use them via Hooks.
+You can use Connections directly from your own code, you can use them via Hooks or use them from :ref:`templates <concepts:jinja-templating>`::
 
+
+    echo {{ conn.<variable_name>.host }}
 
 Hooks
 -----

--- a/docs/apache-airflow/macros-ref.rst
+++ b/docs/apache-airflow/macros-ref.rst
@@ -102,9 +102,8 @@ supplied in case the variable does not exist.
 Similarly, Airflow Connections data can be accessed via the ``conn`` template variable.
 For example, you could use expressions in your templates like ``{{ conn.my_conn_id.login }}``,
 ``{{ conn.my_conn_id.password }}``, etc.
-Just like with ``var`` it's possible to fetch a connection by string or provide defaults with
-``{{ conn.get('my_conn_id').host }}`` or
-``{{ conn.get('my_conn_id', {"host": "host1", "login": "user1"}).host }}``
+Just like with ``var`` it's possible to fetch a connection by string  (e.g. ``{{ conn.get('my_conn_id_'+index).host }}``
+) or provide defaults (e.g ``{{ conn.get('my_conn_id', {"host": "host1", "login": "user1"}).host }}``)
 
 Macros
 ------

--- a/docs/apache-airflow/macros-ref.rst
+++ b/docs/apache-airflow/macros-ref.rst
@@ -69,6 +69,8 @@ Variable                                Description
 ``{{ var.json.my_var.path }}``          global defined variables represented as a dictionary
                                         with deserialized JSON object, append the path to the
                                         key within the JSON object
+``{{ conn.my_conn_id }}``               connection represented as a dictionary
+
 ``{{ task_instance_key_str }}``         a unique, human-readable key to the task instance
                                         formatted ``{dag_id}__{task_id}__{ds_nodash}``
 ``{{ conf }}``                          the full configuration object located at
@@ -96,6 +98,13 @@ It is also possible to fetch a variable by string if needed with
 ``{{ var.value.get('my.var', 'fallback') }}`` or
 ``{{ var.json.get('my.dict.var', {'key1': 'val1'}) }}``. Defaults can be
 supplied in case the variable does not exist.
+
+Similarly, Airflow Connections data can be accessed via the ``conn`` template variable.
+For example, you could use expressions in your templates like ``{{ conn.my_conn_id.login }}``,
+``{{ conn.my_conn_id.password }}``, etc.
+Just like with ``var`` it's possible to fetch a connection by string or provide defaults with
+``{{ conn.get('my_conn_id').host }}`` or
+``{{ conn.get('my_conn_id', {"host": "host1", "login": "user1"}).host }}``
 
 Macros
 ------

--- a/tests/models/test_taskinstance.py
+++ b/tests/models/test_taskinstance.py
@@ -40,6 +40,7 @@ from airflow.exceptions import (
 )
 from airflow.models import (
     DAG,
+    Connection,
     DagRun,
     Pool,
     RenderedTaskInstanceFields,
@@ -59,6 +60,7 @@ from airflow.ti_deps.dependencies_states import RUNNABLE_STATES
 from airflow.ti_deps.deps.base_ti_dep import TIDepStatus
 from airflow.ti_deps.deps.trigger_rule_dep import TriggerRuleDep
 from airflow.utils import timezone
+from airflow.utils.db import merge_conn
 from airflow.utils.session import create_session, provide_session
 from airflow.utils.state import State
 from airflow.utils.types import DagRunType
@@ -67,6 +69,7 @@ from tests.models import DEFAULT_DATE
 from tests.test_utils import db
 from tests.test_utils.asserts import assert_queries_count
 from tests.test_utils.config import conf_vars
+from tests.test_utils.db import clear_db_connections
 
 
 class CallbackWrapper:
@@ -1552,6 +1555,43 @@ class TestTaskInstance(unittest.TestCase):
         assert isinstance(template_context["execution_date"], pendulum.DateTime)
         assert isinstance(template_context["next_execution_date"], pendulum.DateTime)
         assert isinstance(template_context["prev_execution_date"], pendulum.DateTime)
+
+    @parameterized.expand(
+        [
+            ('{{ conn.value.get("a_connection").host }}', 'hostvalue'),
+            ('{{ conn.value.get("a_connection", "unused_fallback").host }}', 'hostvalue'),
+            ('{{ conn.value.get("missing_connection", {"host": "fallback_host"}).host }}', 'fallback_host'),
+            ('{{ conn.value.a_connection.host }}', 'hostvalue'),
+            ('{{ conn.value.a_connection.login }}', 'loginvalue'),
+            ('{{ conn.value.a_connection.password }}', 'passwordvalue'),
+        ]
+    )
+    def test_template_with_connection(self, content, expected_output):
+        """
+        Test the availability of variables in templates
+        """
+        with create_session() as session:
+            clear_db_connections(add_default_connections_back=False)
+            merge_conn(
+                Connection(
+                    conn_id="a_connection",
+                    conn_type="a_type",
+                    description="a_conn_description",
+                    host="hostvalue",
+                    login="loginvalue",
+                    password="passwordvalue",
+                    schema="schemavalues",
+                ),
+                session,
+            )
+
+        with DAG('test-dag', start_date=DEFAULT_DATE):
+            task = DummyOperator(task_id='op1')
+
+        ti = TI(task=task, execution_date=DEFAULT_DATE)
+        context = ti.get_template_context()
+        result = task.render_template(content, context)
+        assert result == expected_output
 
     @parameterized.expand(
         [

--- a/tests/models/test_taskinstance.py
+++ b/tests/models/test_taskinstance.py
@@ -1564,6 +1564,7 @@ class TestTaskInstance(unittest.TestCase):
             ('{{ conn.a_connection.host }}', 'hostvalue'),
             ('{{ conn.a_connection.login }}', 'loginvalue'),
             ('{{ conn.a_connection.password }}', 'passwordvalue'),
+            ('{{ conn.a_connection.extra_dejson["extra__asana__workspace"] }}', 'extra1'),
         ]
     )
     def test_template_with_connection(self, content, expected_output):
@@ -1581,6 +1582,9 @@ class TestTaskInstance(unittest.TestCase):
                     login="loginvalue",
                     password="passwordvalue",
                     schema="schemavalues",
+                    extra={
+                        "extra__asana__workspace": "extra1",
+                    },
                 ),
                 session,
             )

